Invisible layers should not be flagged as separated

etiennesegonzac opened the pull request:
#### 1400c1d4ef81d8cb50b247fc742d86921523cc87
<pre>
Invisible layers should not be flagged as separated
<a href="https://bugs.webkit.org/show_bug.cgi?id=281855">https://bugs.webkit.org/show_bug.cgi?id=281855</a>
&lt;<a href="https://rdar.apple.com/138188635">rdar://138188635</a>&gt;

Reviewed by Simon Fraser.

Separating a layer is not free and only visible layers should get this
treatment.
Check the layer visible rect before applying the `isSeparated` property.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::shouldBeSeparated const):
(WebKit::RemoteLayerTreeNode::setShouldBeSeparated):
Keep track of the requested separated state on the node like we do for
the visible rect.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
Only set `separated` on the layer if its visible

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::updateCoverage):
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
Make sure the visible rect is propagated to the UI Process when
separated layers are enabled.

* LayoutTests/platform/visionos/transforms/separated.html:
* LayoutTests/platform/visionos/transforms/separated-expected.txt:
Update the test with an out-of-viewport separated layer.
* LayoutTests/platform/visionos/transforms/separated-update-expected.txt:
* LayoutTests/platform/visionos/transforms/separated-update.html:
Introduce a new test where scrolling reveals new separated layers.

Canonical link: <a href="https://commits.webkit.org/285567@main">https://commits.webkit.org/285567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecd008df231b455f83c11ca59fcf66fbfe6e7e96

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52474 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25853 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77255 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24283 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60279 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57410 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15898 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76112 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37829 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44064 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20339 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22612 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20694 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78923 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65853 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65129 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16103 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7119 "Passed tests") | | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3061 "Built successfully") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -346,13 +338,29 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
+    if (properties.changedProperties & LayerChange::SeparatedChanged)
+        node.setShouldBeSeparated(properties.isSeparated);
+
+    if (properties.changedProperties & LayerChange::SeparatedChanged || properties.changedProperties & LayerChange::VisibleRectChanged) {
```
The visible rect code in WebCore is gated on `ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)`, so I think you should generalize it to `ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION) | | HAVE(CORE_ANIMATION_SEPARATED_LAYERS)`

etiennesegonzac committed:
Invisible layers should not be flagged as separated
https://bugs.webkit.org/show_bug.cgi?id=281855
<rdar://138188635>

Reviewed by Simon Fraser.

Separating a layer is not free and only visible layers should get this
treatment.
Check the layer visible rect before applying the `isSeparated` property.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::shouldBeSeparated const):
(WebKit::RemoteLayerTreeNode::setShouldBeSeparated):
Keep track of the requested separated state on the node like we do for
the visible rect.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
Only set `separated` on the layer if its visible

* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::updateCoverage):
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
Make sure the visible rect is propagated to the UI Process when
separated layers are enabled.

* LayoutTests/platform/visionos/transforms/separated.html:
* LayoutTests/platform/visionos/transforms/separated-expected.txt:
Update the test with an out-of-viewport separated layer.
* LayoutTests/platform/visionos/transforms/separated-update-expected.txt:
* LayoutTests/platform/visionos/transforms/separated-update.html:
Introduce a new test where scrolling reveals new separated layers.

Canonical link: https://commits.webkit.org/285567@main
